### PR TITLE
Update GNU container recipe, add Clang container

### LINKED ISSUE #163: Need container extra instructions prior to build phase
### Summary

Some containers (e.g. Intel containers) need extra instructions before the build phase, for example adding new `apt` sources and installing the Intel oneAPI suite from there. This should be implemented for both Docker and Singularity using the existing extra-instructions logic.

### Rationale

_No response_

### Description

_No response_

### Additional information

_No response_

### General information

- [X] I have run `spack --version` and reported the version of Spack
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by climbfuji**
## Description

1. Update ubuntu-gnu-openmpi docker recipe:
    - use the default `gnu@9.4.0` compiler instead of `gnu@10.0.0` - there were problems with both of them in the container build
    - use a few more external packages such as `qt@5` and `wget`
    - set environment variable `TZ` to avoid the container build hanging indefinitely in the final phase when installing `tzdata`
    - do not strip binaries after build to avoid warnings at end of build process
    - Use new functionality of extra instructions in pre-build and pre-final phases
2. Add ubuntu-clang-mpich docker recipe, almost identical to ubuntu-gnu-openmpi with minimal necessary changes
3. Add `README.md` in `configs/containers` to list the specs to install for each spack-stack release and application

**Notes**
- CI tests passed for https://github.com/NOAA-EMC/spack-stack/pull/320/commits/5f89cd003c7360c73a09e66f12d9e20610197aaf (this includes installing `py-f90nml` and `py-xarray`)
- Also fixed two small errors in `doc/source/Platforms.rst`
- Waiting on https://github.com/NOAA-EMC/spack/pull/162 (merged)
- Waiting on https://github.com/NOAA-EMC/spack/pull/169 (merged)
- Fixes https://github.com/NOAA-EMC/spack/issues/163

### CODE REVIEW

**Comment by climbfuji** [comment]
Excellent: on the new clang+mpich container, only the folllowing three tests fail in `jedi-bundle`:
```
Total Test time (real) = 9047.02 sec

The following tests FAILED:
	1165 - ufo_test_tier1_test_ufo_conventionalprofileprocessing_UInterp_OPScomparison (Failed)
	1317 - test_soca_convertstate (Failed)
	1375 - test_soca_3dvarlowres_soca (Failed)
Errors while running CTest
Output from these tests are in: /root/jedi/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```

**Comment by climbfuji** [comment]
For jedi-bundle `ctest` failures with gcc+openmpi, see https://github.com/JCSDA-internal/jedi-bundle/issues/12

**Comment by climbfuji** [comment]
@mer-a-o This is definitely ready for testing from your side. See comments in this PR about my testing etc.

File: configs/containers/docker-ubuntu-clang-mpich.yaml [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        RUN cd /usr/bin && \
+        ln -svf clang-10 clang && \
+        ln -svf clang++-10 clang++ && \
+        ln -svf clang-cpp-10 clang-cpp && \
+        cd /usr/lib/llvm-10/lib && \
+        ln -svf libc++abi.so.1.0 libc++abi.so
+      pre_final: |
+        #Set environment variables for installing tzdata
+        ENV DEBIAN_FRONTEND=noninteractive && \
+        TZ=Etc/UTC && \
```
Comment (mer-a-o) [inline configs/containers/docker-ubuntu-clang-mpich.yaml]: This line doesn't need `&& \`. However, after removing `&& \` I still get and error with docker build command complaining about extra `&&` in the Dockerfile. I had remove the space between `&&` and `\` in for it to work, which doesn't make much sense. So with this I didn't get the error about extra `&&`
```
ENV DEBIAN_FRONTEND=noninteractive &&\
TZ=Etc/UTC
```

The error I get:
```
Error response from daemon: failed to parse Dockerfile: Syntax error - can't find = in "&&". Must be of the form: name=value
```

Comment (mer-a-o) [reply configs/containers/docker-ubuntu-clang-mpich.yaml]: ok, I figured it out. You don't need `&&` in ENV. So it should be
```
        ENV DEBIAN_FRONTEND=noninteractive 
        TZ=Etc/UTC
```

File: configs/containers/docker-ubuntu-gcc-openmpi.yaml [hunk 2]
```diff
@@ -68,53 +98,71 @@ spack:
...
+      - python3-pip
+      - python3-setuptools
+      - unzip
+      - vim
 
     # Extra instructions
     extra_instructions:
+      pre_final: |
+        #Set environment variables for installing tzdata
+        ENV DEBIAN_FRONTEND=noninteractive &&\
```
Comment (mer-a-o) [inline configs/containers/docker-ubuntu-gcc-openmpi.yaml]: Please remove `&&\` from here too. Thanks.

**Comment by mer-a-o** [comment]
Building clang container failed with this error:
```
==> Error: ProcessError: Command exited with status 2:
    'make' '-j2'

2 errors found in build log:
     1     ==> pigz: Executing phase: 'edit'
     2     ==> [2022-09-07-19:16:46.811453] Using default implementation: skipp
           ing edit phase.
     3     ==> pigz: Executing phase: 'build'
     4     ==> [2022-09-07-19:16:46.833180] 'make' '-j2'
     5     gcc -O3 -Wall -Wextra -Wno-unknown-pragmas -Wcast-qual   -c -o pigz.
           o pigz.c
     6     gcc -O3 -Wall -Wextra -Wno-unknown-pragmas -Wcast-qual   -c -o yarn.
           o yarn.c
  >> 7     pigz.c:415:10: fatal error: zlib.h: No such file or directory
     8       415 | #include "zlib.h"       // deflateInit2(), deflateReset(), d
           eflate(),
     9           |          ^~~~~~~~
     10    compilation terminated.
  >> 11    make: *** [<builtin>: pigz.o] Error 1
     12    make: *** Waiting for unfinished jobs....

See build log for details:
  /tmp/root/spack-stage/spack-stage-pigz-2.7-fhhp3sk4w6ipgtmjsc7cdinvbpuaqe6x/spack-build-out.txt
```

**Comment by climbfuji** [comment]
> Building clang container failed with this error:
> 
> ```
> ==> Error: ProcessError: Command exited with status 2:
>     'make' '-j2'
> 
> 2 errors found in build log:
>      1     ==> pigz: Executing phase: 'edit'
>      2     ==> [2022-09-07-19:16:46.811453] Using default implementation: skipp
>            ing edit phase.
>      3     ==> pigz: Executing phase: 'build'
>      4     ==> [2022-09-07-19:16:46.833180] 'make' '-j2'
>      5     gcc -O3 -Wall -Wextra -Wno-unknown-pragmas -Wcast-qual   -c -o pigz.
>            o pigz.c
>      6     gcc -O3 -Wall -Wextra -Wno-unknown-pragmas -Wcast-qual   -c -o yarn.
>            o yarn.c
>   >> 7     pigz.c:415:10: fatal error: zlib.h: No such file or directory
>      8       415 | #include "zlib.h"       // deflateInit2(), deflateReset(), d
>            eflate(),
>      9           |          ^~~~~~~~
>      10    compilation terminated.
>   >> 11    make: *** [<builtin>: pigz.o] Error 1
>      12    make: *** Waiting for unfinished jobs....
> 
> See build log for details:
>   /tmp/root/spack-stage/spack-stage-pigz-2.7-fhhp3sk4w6ipgtmjsc7cdinvbpuaqe6x/spack-build-out.txt
> ```

I just fixed a minor bug in the `pre_final`/`final` instructions for the clang container. With those, unrelated to your error, I could build the container just fine. I am using `Docker version 20.10.14, build a224086` on my macOS.

**Comment by climbfuji** [comment]
@mer-a-o This is ready for review.

**Review by mer-a-o — APPROVED** [review]

**Comment by climbfuji** [comment]
I am going to merge this now. Any updates needed will be made in a follow-up PR. Thanks for the review, @mer-a-o!

**Pull request merged**